Add tests for Archive Directory Traversal (#1138)
OWASP/wstg: document/4-Web_Application_Security_Testing/10-Business_Logic_Testing/09-Test_Upload_of_Malicious_Files.md

- In badly configured versions of Nginx, uploading a file as `test.jpg/x.php` may allow it to be executed as `x.php`.

### Malicious File Contents

Once the file type has been validated, it is important to also ensure that the contents of the file are safe. This is significantly harder to do, as the steps required will vary depending on the types of file that are permitted.

#### Malware

Applications should generally scan uploaded files with anti-malware software to ensure that they do not contain anything malicious. The easiest way to test for this is using the [EICAR test file](https://www.eicar.org/download-anti-malware-testfile/), which is an safe file that is flagged as malicious by all anti-malware software.

Depending on the type of application, it may be necessary to test for other dangerous file types, such as Office documents containing malicious macros. Tools such as the [Metasploit Framework](https://github.com/rapid7/metasploit-framework) and the [Social Engineer Toolkit (SET)](https://github.com/trustedsec/social-engineer-toolkit) can be used to generate malicious files for various formats.

> When this file is uploaded, it should be detected and quarantined or deleted by the  application. Depending on how the application processes the file, it may not be obvious whether this has taken place.

Reviewer: Still

#### Archive Directory Traversal

> If the application extracts archives (such as ZIP files), then it may be possible to write to unintended locations using directory traversal. This can be exploited by uploading a malicious ZIP file that contains paths that traverse the file system using sequences such as `..\..\..\..\shell.php`. This technique is discussed further in the [snyk advisory](https://snyk.io/research/zip-slip-vulnerability).

Reviewer 2: We already have a reference to the paper in the references section at the bottom, so I don't really think that we need another one here? Or if this is a better link, then replace the other reference with it.

> If the application extracts archives (such as ZIP files), then it may be possible to write to unintended locations using directory traversal. This can be exploited by uploading a malicious ZIP file that contains paths that traverse the file system using sequences such as `..\..\..\..\shell.php`. This technique is discussed further in the [snyk advisory](https://snyk.io/research/ZIP-slip-vulnerability).

Author: Reverted!

A test against Archive Directory Traversal should include two parts:

> 1. A malicious archive that breaks out of the target directory when extracted. This malicious archive can contain two files: a 'notinfected.sh' file, extracted into the target directory, and also an 'infected.sh' file, that attempts to navigate up the directory tree to hit the root folder - adding a file into the tmp directory. A malicious path can contain many levels of '../' (i.e. ../../../../../../../../tmp/infected.sh) to stand a better chance of reaching the root directory.

Reviewer 2: Including the two separate files in a directory is a good way to test this - but I wonder if `infected.sh` is a sensible filename? We don't need it to be a `.sh` file to prove the concept, and a sysadmin finding a file called `infected.sh` on their system is likely to cause unintended issues.

The two files could just as easily be called `base` and `traversal` or something more neutral, that's less likely to cause a panic.

Author: It makes sense to me and I renamed them.

> 1. A malicious archive, that extracted breaks out of the target directory. A compressed file could contains two files: a “notinfected.sh” file, extracted into the target directory, and a “infected.sh” file, that attempts to navigate to the directory tree to hit the root folder - adding a file into the tmp directory. A malicious path could contain many levels of "../" (i.e. ../../../../../../../../tmp/infected.sh) to stand a better chance of reaching the root directory.

Reviewer: The text needs a tweak too

> 2. A functionality, that is required to extract compressed files, either using custom code or a library. Archive Directory Traversal vulnerability exists when the extraction functionality don’t validate file paths in the archive. The example below shows a vulnerable implementation in Java:

Author: Done!

> 2. A functionality, that is required to extract compressed files, either using custom code or a library. Archive Directory Traversal vulnerability exists when the extraction functionality don’t validate file paths in the archive. The example below shows a vulnerable implementation in Java:

Reviewer: don't ↔️ doesn't

```java
    Enumeration<ZipEntry>entries=​​zip​.g​etEntries(); 
        while(entries​.h​asMoreElements()){

>             ZipEntry e ​=​entries.nextElement();

Reviewer 2: The space before the `=` but not after looks weird to me - is this a Java thing?

            File f = new File(destinationDir, e.getName());
            InputStream input =zip​.g​etInputStream(e);
            IOUtils​.c​opy(input, write(f));
        }
```

Additional testing techniques:

> - Upload a malicious ZIP file and try to remote access this file when upload is completed.  [Watch it in action here](https://www.youtube.com/watch?v=l1MT5lr4p9o)

Reviewer 2: I don't really see what a YouTube video adds to this? If the explanation in the WSTG isn't detailed enough then we should expand it - it's not use if I'm trying to test for this vulnerability in an environment where I don't have Internet access and the WSTG just says "go on YouTubeand watch this video".

So something like

* Create a ZIP containing two files - with with a traversed path such as `../../traversed` and one without `base`.
* Upload the file to the application.
* Verify that the two files are located within different folders on the web server after the archive has been extracted.

It would also be useful to provide some guidance on how you actually create a ZIP archive to use for this.

> - Include a unit test to upload an infected compressed file on the extraction method.

Reviewer: ```suggestion
- Include a unit test to upload an infected compressed file then execute the extraction method.
```

Author: Done!

> - Validate that libraries being used have been [patched for this vulnerability.](https://github.com/snyk/ZIP-slip-vulnerability#affected-libraries)

Reviewer: ZIP in the URL can be lower cased. (If the bot still complains I'll deal with it)

Author: Reverted!

- Include a validation that throws an exception when vulnerabilities is included, like in the example below:

```java
    StringcanonicalDestinationDirPath=destinationDir.getCanonicalPath();
    Filedestinationfile=newFile(destinationDir,e.getName());
    StringcanonicalDestinationFile=destinationfile.getCanonicalPath();
    if(!canonicalDestinationFile.startsWith(canonicalDestinationDirPath+File.separator)){
        throw new ArchiverException("Entry is outside of the target dir: " + e.getName()); 

>     }

Reviewer 2: This isn't really a "testing technique", and seems a bit too specific to in the WSTG as a recommendation.

Author: @rbsec  You are right, after reviewing it it looks more a remediation. Should we move it to remediation section instead?

Reviewer 2: @doverh most of the other remediation sections in the WSTG are more general - usually a few bullet points giving a high level explanation of how to fix the issue, rather than specific code samples. This means that they're more generally applicable, whereas this kind of code sample is only really useful if they're using the specific language and framework of that code.

So perhaps taking out the code block, but giving a brief explanation of what they should do to fix it in the remediation section would be better?

Author: @rbsec I reviewed File Upload Cheat Sheet  and it brings sections with instructions on File Upload Protection and Filename Sanitization that present a general level explanation for the Java code I share as vulnerable. Do you believe this is adequate?

Reviewer 2: If we can link to something like a cheat sheet in the recommendations that's always good.

> ```

Reviewer: Again probably doesn't need line numbers.

Author: Line number removed!

2. Logic that extracts compressed files either using custom code or a library. Archive Directory Traversal vulnerabilities exist when the extraction functionality doesn’t validate file paths in the archive. The example below shows a vulnerable implementation in Java:

```java
Enumeration<ZipEntry> entries =​ ​zip​.g​etEntries();

while(entries​.h​asMoreElements()){
    ZipEntry e ​= ​entries.nextElement();
    File f = new File(destinationDir, e.getName());
    InputStream input = zip​.g​etInputStream(e);
    IOUtils​.c​opy(input, write(f));
}
```

Follow the steps below to create a ZIP file that can abuse the vulnerable code above once its uploaded to the web server:

```bash
# Open a new terminal and create a tree structure
# (more directory levels might be required based on the system being targeted)
mkdir -p a/b/c
# Create a base file
echo 'base' > a/b/c/base
# Create a traversed file
echo 'traversed' > traversed
# You can double check the tree structure using `tree` at this stage
# Navigate to a/b/c root directory
cd a/b/c
# Compress the files
zip test.zip base ../../../traversed
# Verify compressed files content

> nzip -l test.zip

Reviewer 2: Typo, should be `unzip`

Reviewer 3: I think after that line there's a missing triple quote as well

Reviewer: Ya there should be triple backticks on the next line to end the bash code fence.

1. A malicious archive that breaks out of the target directory when extracted. This malicious archive can contain two files: a 'base' file, extracted into the target directory, and also an 'traversed' file, that attempts to navigate up the directory tree to hit the root folder - adding a file into the tmp directory. A malicious path can contain many levels of '../' (i.e. ../../../../../../../../tmp/traversed) to stand a better chance of reaching the root directory.
2. A functionality, that is required to extract compressed files, either using custom code or a library. Archive Directory Traversal vulnerabilities exist when the extraction functionality doesn’t validate file paths in the archive. The example below shows a vulnerable implementation in Java:

```java
    Enumeration<ZipEntry> entries =​ ​zip​.g​etEntries(); 
        while(entries​.h​asMoreElements()){
            ZipEntry e ​= ​entries.nextElement();
            File f = new File(destinationDir, e.getName());
            InputStream input = zip​.g​etInputStream(e);
            IOUtils​.c​opy(input, write(f));
        }
```

Additional testing techniques:

> - Upload a malicious ZIP file and try to remote access this file when upload is completed.
>     1. Open a new terminal and create tree structure:
>     Note: more or less levels of directories are required depending on how far you need to traverse.
>         `~$ mkdir -p a/b/c`
>     2. Create a base file:
>         `~$ echo base > a/b/c/base`
>     3. Create a traversed file:
>         `~$ echo traversed > traversed`
>     4. Validate your tree structure:
>         `~$ tree`
>     5. Navigate to abc root directory:
>         `~$ cd a/b/c`
>     6. Compress files:
>     `~$ zip test.zip base ../../../traversed`
>     7.Verify compressed files content:
>     `~$ nzip -l test.zip`

Reviewer 3: Gave it a small clean up and transformed into one script that can be followed through the comments, as currently it's getting badly transformed into markdown.
```suggestion
```bash
# Open a new terminal and create a tree structure
# (more directory levels might be required based on the system being targeted)
mkdir -p a/b/c
# Create a base file
echo 'base' > a/b/c/base
# Create a traversed file
echo 'traversed' > traversed
# You can double check the tree structure using `tree` at this stage
# Navigate to a/b/c root directory
cd a/b/c
# Compress the files
zip test.zip base ../../../traversed
# Verify compressed files content
nzip -l test.zip
```
```

> - Include a unit test to upload an infected compressed file then execute the extraction method.
> - Validate that libraries being used have been [patched for this vulnerability.](https://github.com/snyk/zip-slip-vulnerability#affected-libraries)

Reviewer 3: We recommend these bits across the book as a whole, and this can be applied to all the test cases, whether it's discussing a SCA integration (so Snyk as you reference it), or through unit tests, so we don't need to mention it within a subcategory of a test file.
```suggestion
```

A test against Archive Directory Traversal should include two parts:

1. A malicious archive that breaks out of the target directory when extracted. This malicious archive can contain two files: a 'base' file, extracted into the target directory, and also an 'traversed' file, that attempts to navigate up the directory tree to hit the root folder - adding a file into the tmp directory. A malicious path can contain many levels of '../' (i.e. ../../../../../../../../tmp/traversed) to stand a better chance of reaching the root directory.
2. A functionality, that is required to extract compressed files, either using custom code or a library. Archive Directory Traversal vulnerabilities exist when the extraction functionality doesn’t validate file paths in the archive. The example below shows a vulnerable implementation in Java:

```java

>     Enumeration<ZipEntry>entries=​​zip​.g​etEntries(); 

Reviewer: ```suggestion
    Enumeration<ZipEntry> entries =​ ​zip​.g​etEntries(); 
```

        while(entries​.h​asMoreElements()){
            ZipEntry e ​= ​entries.nextElement();
            File f = new File(destinationDir, e.getName());
            InputStream input = zip​.g​etInputStream(e);
            IOUtils​.c​opy(input, write(f));
        }
```

Additional testing techniques:

- Upload a malicious ZIP file and try to remote access this file when upload is completed.
    1. Open a new terminal and create a new folder:

>         'mkdir <zip_folder>'

Reviewer: ```suggestion
        `mkdir <zip_folder>`
```

    2. Create a base file:

>         'touch base.txt'

Reviewer: ```suggestion
        `touch base.txt`
```

    3. Open this file, add a simple note and save it.
    4. Create a traversed file that matches a local or remote directory:

>         'touch ../../../../../../../../tmp/traversed'

Reviewer: ```suggestion
        `touch ../../../../../../../../tmp/traversed`
```

    5. Open this file and a message to echo (executing this file should echo this message):

>         echo 'Your message here'

Reviewer: ```suggestion
        `echo 'Your message here'`
```

>         echo 'Your message here'

Reviewer: ```suggestion
        `echo 'Your message here'`
```

    6. Create the ZIP file:

>         'ZIP -r <zip_filename> <directoty_name>'

Reviewer: ```suggestion
        'zip -r <zip_filename> <directoty_name>'
```

>         'zip -r <zip_filename> <directoty_name>'

Reviewer: ```suggestion
        `zip -r <zip_filename> <directoty_name>`
```

>         `zip -r <zip_filename> <directoty_name>`

Reviewer 2: I don't think these instructions actually work, because you're just creating a zip of the folder here, rather than the file and a traversed path. I think you need to do something like this:

```
~$ mkdir -p a/b/c
~$ echo base > a/b/c/base
~$ echo traversed > traversed
~$ tree
.
├── a
│   └── b
│       └── c
│           └── base
└── traversed

~$ cd a/b/c
~/a/b/c$ zip test.zip base ../../../traversed
  adding: base (stored 0%)
  adding: ../../../traversed (stored 0%)

And then to verify:

~/a/b/c$ unzip -l test.zip
Archive:  test.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
        5  2024-04-24 10:30   base
       10  2024-04-24 10:30   ../../../traversed
---------                     -------
       15                     2 files
```

Perhaps with a note that you may need more or less levels of directories, depending on how far you need to traverse. And since we're using `zip` to create the archive, it seems sensible to use `unzip` to verify it rather than `jar`. The tree command isn't needed (and could perhaps be removed) - it's more just to show the structure we're creating.

    7. Validate files compressed

>         'jar -tvf <zip_filename>'

Reviewer: ```suggestion
        `jar -tvf <zip_filename>`
```

    4. Create a traversed file that matches a local or remote directory:
        touch ../../../../../../../../tmp/traversed
    5. Open this file and a message to echo (executing this file should echo this message):
        echo 'Your message here'
    6. Create the zip file:
        zip -r 'zip file name' 'directory name
    7. Validate files compressed

>         jar -tvf 'zip file name'

Reviewer: You could use inline code fences if that makes more sense. (Single backtick before and after)

Reviewer: Like:
&#96;`jar -tvf <zip file name>`&#96;

- [CWE-434: Unrestricted Upload of File with Dangerous Type](https://cwe.mitre.org/data/definitions/434.html)
- [Implementing Secure File Upload](https://infosecauditor.wordpress.com/tag/malicious-file-upload/)
- [Metasploit Generating Payloads](https://www.offensive-security.com/metasploit-unleashed/Generating_Payloads)

> - [ZIP Slip](https://res.cloudinary.com/snyk/image/upload/v1528192501/ZIP-slip-vulnerability/technical-whitepaper.pdf)

Reviewer: In the URL can likely be lower cased.

Author: Done!
